### PR TITLE
Food Health System V2: Merge-conflict Boogaloo

### PR DESCRIPTION
**Pull request by sawu-tg**
- Added a "Food Health" system. The closer a food is to it's wanted temp, the 'Healthier" it is
- Foods that are left out will begin to rot and become toxic
- Added some configurable serving stations to the Kitchen that warm and cool food
- Refrigerators will instantly cool food.
- Chef has been given a "smart" food-train to promote rotation of foodstuffs (no pun intended)
  sushitrain
  kitchen2
  kitchen3

### CODE REVIEW

**Comment by S0ldi3rKr4s0** [comment]
>only fahrenheit

**Comment by sawu-tg** [comment]
@S0ldi3rKr4s0 > Multiplying the Temperature value by 15.5 to achieve a fake-but-real temperature

**Comment by paprka** [comment]
>people actually use centigrade in 2015

**Comment by WJohn** [comment]
Yeah every other form of temp in the game is either centigrade or kelvin.

**Comment by paprka** [comment]
yes but americans = food = fahrenheit

chegmate

**Comment by sawu-tg** [comment]
@paprka I'd like to [remind you](http://i.imgur.com/3ZidINK.png?1) that 90% of the world uses Celsius, I just went with Fahrenheit because Nanotrasen seems very american. (also fuck working out a number to fake celsius)

**Comment by AlexanderUlanH** [comment]
If this is going to affect mob temperature, it'll probably have to be kelvin anyway.
Plus, it would be better fluff-wise to make ingame temperature measurements as consistent as possible.

**Comment by sawu-tg** [comment]
@AlexanderUlanH It does not affect mob temperature. The temperature (in Fahrenheit) is entirely fluff, the temperature variable is how fast it shifts from one to the other. The only way it will affect a mob is if it is too hot or cold, in which case it burns them.

If everyone wants kelvin instead, then I will change it, but I require the appropriate conversation (considering I pulled the 15.5 out of thin air)

**Comment by ike709** [comment]
Let's use kelvin because I think chemistry does. CONSISTENCY MATTERS

**Comment by Jordie0608** [comment]
Kelvin is rather obtuse when dealing with food, centigrade makes more sense.

If i'm not mistaken from glancing at the code, changing the multiplier to 10 should make the highest normal temp to 100, boiling temp in centigrade, which is beyond food edibility temperature and would burn you.

If the multiplier was 7-8 then the max normal temp would be 70-80 centigrade, which is generally how hot food should be cooked for it to be edible.

**Comment by AlexCrimson** [comment]
This seems like a massive pain.

**Comment by sawu-tg** [comment]
@AlexCrimson I made it for three reasons: 1) A replacement/standardisation of the donk pocket's unique reagents that people whined about being removed, 2) To encourage proper handling and use of food, and 3) To give the chef more things to do, beyond being a food conveyor.

In the long run? For lazy people it means two seconds of setting up table temps, and then remembering which table is which. For regular players, it means more depth to the way they serve and handle food.

**Comment by AlexCrimson** [comment]
How do we know its wanted temperature? I mean for Chefs im sure it will be easy, but people are just going to grab food from the stands regardless as to what temperature its at. Unless there is some very obvious visual indicator that its too hot/cold to eat.

Once again, it just seems like a needlessly complicated feature that will just result in people complaining about toxic food. The code also says something about emagging the serving stations? How is a chef supposed to fix them?

**Comment by Alek2ander** [comment]
It would make more sense to use a microwave to heat and a fridge to cool stuff.

**Comment by sawu-tg** [comment]
@AlexCrimson Foods only need to be hot or cold to be preserved, and as the pictures show, their name is in red (for foods needing warming) and blue (for foods needing chilling). It also shows the temperature of each piece of food, so people who check beforehand can see if it's at a right temperature, there will always be people who assume or just take things, and those are the people that get angry when they eat high potency chilli soup without bothering to look at it.

Like a lot of things that have emag_act(), there is no repair for them.

@Alek2ander (Smart)fridges cool, and Microwaves heat, both to their respective maximum, but the food does not stay hot/cold afterwards (it returns to "room temperature"). The serving plates allow food to maintain a constant temperature, a la buffets.

File: code/modules/food&drinks/kitchen machinery/servingplate.dm [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+			if(emagged && !F.tainted)
+				F.tainted = TRUE
+			use_power(active_power_usage * targetTemp)
+
+/obj/machinery/servingplate/proc/fluffTemp(var/what)
+	if(what < 0)
+		what = -what
+	switch(what) //to is not inclusive, needs to be x-1 to y-1
+		if(-110 to 3)
+			return cold ? "Cool" : "Lukewarm"
```
Comment (Miauw62) [inline code/modules/food&drinks/kitchen machinery/servingplate.dm]: if you have something thats boiling and set it to cold, won't it instantly become freezing?
I don't see why you can't just do a straight temperature var.

**Comment by phil235** [comment]
Replacing the tables of the kitchen front with serving plates means people can no longer get in by table climbing.

**Comment by Dorsisdwarf** [comment]
Realisms detected. Activate the danno.

File: code/modules/food&drinks/kitchen machinery/servingplate.dm (hunk 1 above)
Comment (sawu-tg) [inline code/modules/food&drinks/kitchen machinery/servingplate.dm]: @Miauw62 No. It instantly says "boiling/chilling" when you switch, but it takes a moment for the temperature to go down or up. Cold is a negative number, hot is a positive.

If that's too confusing I can change it.

Comment (Miauw62) [inline code/modules/food&drinks/kitchen machinery/servingplate.dm]: I just don't see why the cold check here is neccesary.

**Comment by dannno** [comment]
I'm actually not too salty about this
Rotting food sounds kinda shitty IMO but the "healthier" food stuff sounds pretty good
if "healthier" food had more vitamin in it or more of the food-specific reagent some foods have in it, the chef job could go from input/output machine to something that people can actually boast prowess about

**Comment by Razharas** [comment]
Seems interesting

File: code/modules/food&drinks/drinks/drinks/bottle.dm [hunk 2]
```diff
@@ -33,7 +33,7 @@
 	qdel(src)
 
 /obj/item/weapon/reagent_containers/food/drinks/bottle/attack(mob/living/target as mob, mob/living/user as mob)
-
+	..()
```
Comment (tkdrg) [inline code/modules/food&drinks/drinks/drinks/bottle.dm]: This is going to make breaking bottles on someone's head impossible. Why are you adding this?

File: code/modules/food&drinks/drinks/drinks.dm [hunk 3]
```diff
@@ -24,7 +24,7 @@
 	return
 
 /obj/item/weapon/reagent_containers/food/drinks/attack(mob/M as mob, mob/user as mob, def_zone)
-
+	..()
```
Comment (tkdrg) [inline code/modules/food&drinks/drinks/drinks.dm]: Why do you need this here?

File: code/game/objects/structures/crates_lockers/closets/secure/freezer.dm [hunk 4]
```diff
@@ -12,6 +12,14 @@
...
 			icon_state = icon_opened
 
+/obj/structure/closet/secure_closet/freezer/close()
+	..()
+	if(!broken)
+		for(var/O in src)
+			if(istype(O,/obj/item/weapon/reagent_containers/food))
+				var/obj/item/weapon/reagent_containers/food/F = O
+				F.freshMod = -1
+
```
Comment (tkdrg) [inline code/game/objects/structures/crates_lockers/closets/secure/freezer.dm]: Won't this mean that the best way to cool food is spam-clicking the freezer to keep opening/closing its door?

Comment (sawu-tg) [inline code/game/objects/structures/crates_lockers/closets/secure/freezer.dm]: No. It grows/lose temp to achieve 0 if not being affected, meaning that leaving it in the freezer is best. The moment it leaves, it begins to warm.

File: code/modules/food&drinks/drinks/drinks/bottle.dm (hunk 2 above)
Comment (sawu-tg) [inline code/modules/food&drinks/drinks/drinks/bottle.dm]: Not as far as I know. It calls the parent's attack proc, which contains the code to deal with heated items.

File: code/modules/food&drinks/drinks/drinks.dm (hunk 3 above)
Comment (sawu-tg) [inline code/modules/food&drinks/drinks/drinks.dm]: It calls the parent's attack proc, which contains the code to deal with heated items.

**Comment by TheVekter** [comment]
As someone who plays Cook quite often, I'm not a huge fan of these changes, and here's why.

* Causing food to rot needlessly complicates the Cook's job. Instead of him just making food and the station picking it up when it's necessary, he'll need to make everything to order to prevent people dying from toxic food. Given the station's tendency to be super impatient, this is just going to result in more people breaking in to the kitchen to do it themselves. I've seen this happen one too many times when I got caught up in admin stuff and a player decides I'm not making their super bite burger quite as fast as they want.

* I like the concept of keeping foods at temp, but I think the conveyor is laid out poorly. It looks off compared to everything else in the room. We talked about this in coderbus last night, can the belt get a snowflake sprite so it doesn't look so massive? I think I'd like it a lot more if it didn't look like that.

* Is there any reason why there can't just be serving plates out in front instead of the conveyor? It seems like it's needlessly complicating things.

But yeah, remove the food rotting thing and I'm :100: on this one.

**Pull request closed without merging**